### ISSUE TITLE
[NEW] Support for wasm besides LUA as a scripting engine

### ISSUE THREAD

**Issue by andyxning** [madolson/placeholderkv#387]
**The problem/use-case that the feature addresses**
WASM is becoming more and more popular and many programming languages support WASM as an target, such as rust, golang etc. These programming languages are more used by programmers than LUA.
A description of the problem that the feature will solve, or the use-case with which the feature will be used.

**Description of the feature**

A description of what you want to happen.

**Alternatives you've considered**

Any alternative solutions or features you've considered, including references to existing open and closed feature requests in this repository.

**Additional information**

Any additional information that is relevant to the feature request.

**Comment by rjd15372** [comment]
@madolson @zuiderkwast I can start doing some research for this issue if you think it's worth to pursue.

**Comment by madolson** [comment]
@rjd15372 Go for it! I think there is some potential for a more performant alternative to lua.

**Comment by rjd15372** [comment]
Cool, can someone assign me to this issue?

**Comment by andyxning** [comment]
I am also interested in this. Ping me if you have any idea

**Comment by ahmadbelb** [comment]
I am also interested in this, maybe we can collaborate !

**Comment by rjd15372** [comment]
I'm doing the ground work to allow running WASM code in Valkey. I'm pretty sure that there will be a lot of things to do on top of it after the basic functionality is merged into the unstable branch.

**Comment by rjd15372** [comment]
The WASM runtime engine that I'll be using is https://github.com/bytecodealliance/wasmtime

**Comment by zuiderkwast** [comment]
> The WASM runtime engine that I'll be using is https://github.com/bytecodealliance/wasmtime

OK! Will you make it an optional build dependency? It's a different language (Rust) and a different license (Apache 2).

**Comment by andyxning** [comment]
wasm engine does not have much influence on this. we can just run through the functionality first and benchmark it later.

**Comment by andyxning** [comment]
also, any proposal for implementing or architecture this support, i think we need first also make a review on it.

**Comment by rjd15372** [comment]
> 
> OK! Will you make it an optional build dependency? It's a different language (Rust) and a different license (Apache 2).

Yes, the build dependency will be optional. And the idea is to download the pre-built wasm library at valkey build time, so no need to have rust toolchain to build the wasm library.

We can link Valkey against the wasmtime static library, and therefore there's no need to package the wasmtime library along with the Valkey package.

**Comment by rjd15372** [comment]
@zuiderkwast @andyxning I already have a draft commit with the changes to the build scripts to add the wasmtime dependency here: https://github.com/rjd15372/valkey/commit/28848ed12f2901a8295da4cae6756f968f06c171

**Comment by zuiderkwast** [comment]
> And the idea is to download the pre-built wasm library at valkey build time, so no need to have rust toolchain to build the wasm library.

I guess it's handy for some users, but it's completely unacceptable to others, such as the company where I work. Downloading binaries from the internet is completely out of the question. Even downloading source code deps automatically is something frowned upon. Each dependency needs to be approved by legal and security people, etc.

Distros wouldn't want this either. Deb packages depend on other deb packages, RPMs on other RPMs, etc. from the same distro.

So, I'd say let's use the system installed dependency by default, like we do with OpenSSL, and if the location isn't right, the user can specify it as a parameter to `make`.

```
ifeq ($(uname_M),arm64)
	# Homebrew arm64 uses /opt/homebrew as HOMEBREW_PREFIX
	OPENSSL_PREFIX?=/opt/homebrew/opt/openssl
else
	# Homebrew x86/ppc uses /usr/local as HOMEBREW_PREFIX
	OPENSSL_PREFIX?=/usr/local/opt/openssl
endif
```

We could provide a way to download the binary too, but I think it needs to be guarded by a very explicit command, such as `make download-wasm-binary-deps` or so. And possibly do some SHA checksum to verify that the downloaded binary is the expected one. (I've learnt to be paranoid about these things known as supply chain attacks.)

> We can link Valkey against the wasmtime static library, and therefore there's no need to package the wasmtime library along with the Valkey package.

In our binaries we releases, containers? Sure, if it doesn't make our binary too huge.

**Comment by madolson** [comment]
A lot of the foundational work was already developed to allow third party modules to create new scripting engines. The pathway to adding a WASM engine might be to have it exist as a first-party that is optional, which might make it a bit simpler. It also allows us to pull it out of the core and iterate more independently. One thing that's annoying about LUA is we've been stuck with a very old API since we don't want to break scripts.

**Comment by rjd15372** [comment]
@zuiderkwast 
> I guess it's handy for some users, but it's completely unacceptable to others, such as the company where I work. Downloading binaries from the internet is completely out of the question. Even downloading source code deps automatically is something frowned upon. Each dependency needs to be approved by legal and security people, etc.

This a great point, and most linux distributions forbid downloading any dependencies during the packaging process.

As you said, the solution would be to detect if the dependency is installed locally, and if not then allow the developer to choose what to do. Either download the binary using a make explicit command, or allow him to build the dependency from source.

**Comment by rjd15372** [comment]
@madolson 
> A lot of the foundational work was already developed to allow third party modules to create new scripting engines. The pathway to adding a WASM engine might be to have it exist as a first-party that is optional, which might make it a bit simpler. [...]

I did not fully understood your point. Are you saying that a preferable approach would be to add WASM support as an external valkey module?

**Comment by rjd15372** [comment]
@zuiderkwast @andyxning here's a draft PR for the makefile changes https://github.com/valkey-io/valkey/pull/1183

**Comment by zuiderkwast** [comment]
Yeah, it looks like a good start. We can review the details when you have some working implementation.

I believe Madelyn meant that it can be implemented as a module using the module API, yes. There was some effort to do that and I believe Redis Enterprise already used it to support JavaScript for functions and scripts. If the API already provides the integration, it may be beneficial to use it rather than to develop a new way to plug in scripting engines.

So, it'd be good if you could take a look at what can be done with the module API. Even if it's a module, we can decide to include it in the same repo, or to distribute it in some "extended valkey" release with a bunch of additional stuff such as the bloom filter module.

**Comment by andyxning** [comment]
I agree with what @zuiderkwast said above, we need to first write a proposal and have some time to review the implementation plan incase current what has been coded is not the best or preferred way to do this. That's why i suggest above that we need to first make a proposal doc.

**Comment by rjd15372** [comment]
Sure, I'll take a look at the module API and will write something up and post it here.

**Comment by rjd15372** [comment]
We should have two main components to implement WASM support:
* A WASM Valkey Runtime based on wasmtime.
* A WASM Valkey API, that developers can call within the WASM scripts/functions.

The WASM runtime should accept WASM programs written in WAT format, or WASM binary (compiled program). The support for WASM programs written in several languages will depend on the availability of the WASM Valkey API for each language.

The WASM support should provide the same level of functionality as the Lua scripts/functions. The WASM API should provide the same functions and globals as the Lua API.

WASM programs can be registered as scripts, or functions (similarly with the current Lua scripts or functions).

When registering a WASM script, the WASM program should define a single exported function, which our WASM runtime will run.

When registering a WASM function, the WASM program may define multiple exported functions. These functions can then be called later by a similar `FCALL` command.

**Comment by andyxning** [comment]
> The WASM support should provide the same level of functionality as the Lua scripts/functions.

Maybe not. WASM and LUA engine are different runtime engines, let them be different and have different features are preferred, i think.

**Comment by zuiderkwast** [comment]
Providing the same API as for Lua is a good start. It's very small. EVAL just provides `call` to call Valkey commands. FCALL functions add a few more.

I believe EVAL can support multiple languages by including a shebang-like first line, so we can allow EVAL scripts written in WASM.

One of the main powers of WASM is that it's **sandboxed**.

Lua isn't sandboxed by design. We added sandboxing by patching Lua, but it isn't a perfect solution. I wonder if it would be possible to run Lua under WASM to make it truly sandboxed? I found this blog about compiling the Lua VM to WASM: https://twdev.blog/2024/01/wasm_cpp_05/

The performance and sandboxing are great features. In the future we can think of extending the API to use cases that today can only be done in the module API.

**Comment by hwware** [comment]
@andyxning How do you think you can first create a RFC for WASM, so that we can first decide the high level design, including WASM how to intergrate with Valkey core part.  RFC link is here https://github.com/valkey-io/valkey-rfc

**Comment by hwware** [comment]
> Providing the same API as for Lua is a good start. It's very small. EVAL just provides `call` to call Valkey commands. FCALL functions add a few more.
> 
> I believe EVAL can support multiple languages by including a shebang-like first line, so we can allow EVAL scripts written in WASM.
> 
> One of the main powers of WASM is that it's **sandboxed**.
> 
> Lua isn't sandboxed by design. We added sandboxing by patching Lua, but it isn't a perfect solution. I wonder if it would be possible to run Lua under WASM to make it truly sandboxed? I found this blog about compiling the Lua VM to WASM: https://twdev.blog/2024/01/wasm_cpp_05/
> 
> The performance and sandboxing are great features. In the future we can think of extending the API to use cases that today can only be done in the module API.

I think if we import WASM into Valkey, then we wish any programming language is allowed to work with Valkey, thus existing Lua script,  Functions, and WASM are three options for Valkey users

**Comment by andyxning** [comment]
> How do you think you can first create a RFC for WASM, so that we can first decide the high level design, including WASM how to intergrate with Valkey core part. RFC link is here https://github.com/valkey-io/valkey-rfc

@hwware yes, i can do this asap. Pls wait for some days.

**Comment by migueleliasweb** [comment]
Hi all, any updates on this?

Adding wasm support would be great 😃👍.

**Comment by rjd15372** [comment]
In 8.1 release was introduced new module API functions to support the development of new scripting engines as external Valkey modules.

Building a WASM engine module is in the plans but the development hasn't started yet.

**Comment by madolson** [comment]
> Building a WASM engine module is in the plans but the development hasn't started yet.

Contributions are welcome though, we'll help you get started :)

**Comment by jeffparsons** [comment]
Note for anyone who ends up working on this: I would suggest using the Wasm Component Model, rather than defining your own ABI.

**Comment by jeffparsons** [comment]
What's the best place to write up a concrete implementation proposal for this? Just in comments in this issue? In a discussion?

EDIT: Ha, I just saw https://github.com/valkey-io/valkey/issues/3172. :)

**Comment by zuiderkwast** [comment]
@jeffparsons Great, you seem to know stuff in this area. Do you want to create a PoC? I don't care exactly how we communicate, as long as we communicate in some way. :) 

@rjd15372 has done a lot of work to allow implementing scripting engines entirely in the module API. It means it can be done as a separate module, separate repo, and avoid bloating the core repo with these dependencies.

**Comment by jeffparsons** [comment]
I've started hacking on a PoC here: https://github.com/jeffparsons/valkey-wasm. Nothing worth showing off yet — just declaring intent.

Note: all the actual code is written by Claude. I know that a lot of folks have strong feelings about that sort of thing, so I just wanted to get it out there up-front. This is the only realistic path for me these days because I don't have the discretionary time available for writing artisanal hand-crafted code. 😅 

Design decisions of note for the PoC:

- Guests are Wasm Components (no raw modules)
- Only focusing on an EVAL-alike interface for now. (Whole Valkey modules would come later, with an entirely different interface.)
- Initially only a read-only interface. Component instances that only use this interface can be killed off at any time, and _will_ be killed off via fuel metering and/or epoch-based limiting.
- Eventually I plan to extend this to include an interface that allows writing to the cache, but only by accumulating mutating commands into a buffer that will be flushed at the end. In that way, the component instance can _still_ be terminated at any time safely, and the only uninterruptible part is flushing those write commands at the end, which is (a) controlled by the host, and (b) completely deterministic — no getting stuck in loops or anything.
- _Maybe_ eventually a "read-and-write at will" interface equivalent to the current Lua scripting, but this feels like a misfeature to me. My gut says we should be trying to build something _better_ and safer than the existing Lua scripting. I'm still learning about all this (I wrote my first Lua scripts for Valkey a couple of days ago, which is how I ended up here) so my opinions will probably continue to evolve as I hack + learn.

**Comment by rjd15372** [comment]
I did a quick look at the code. It looks like you are not using the scripting engine module API to integrate the WASM engine into Valkey. Are you planning to run the WASM script using a `WASM.EVAL` command?

**Comment by jeffparsons** [comment]
> I did a quick look at the code. It looks like you are not using the scripting engine module API to integrate the WASM engine into Valkey. Are you planning to run the WASM script using a `WASM.EVAL` command?

Oh, right. I missed that this is a separate thing. I will use it. Thanks for the heads-up!
